compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -838,7 +863,10 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                             }
                         }
                         ProjectAndUnifyResult::FailedNormalization => Ok(EvaluatedToAmbig),
-                        ProjectAndUnifyResult::Recursive => Ok(EvaluatedToRecur),
+                        ProjectAndUnifyResult::Recursive => match self.treat_inductive_cycle {
+                            TreatInductiveCycleAs::Ambig => return Ok(EvaluatedToUnknown),
+                            TreatInductiveCycleAs::Recur => return Ok(EvaluatedToRecur),
+                        },
```

Comment:
supr optional nit: have an `Into` impl or a method for conversion.
```suggestion
                        ProjectAndUnifyResult::Recursive => Ok(self.treat_inductive_cycle.into()),
```